mmdet/models/detectors/single_stage_instance_seg.py

```diff
@@ -0,0 +1,168 @@
+        self.test_cfg = test_cfg
+
+    def extract_feat(self, img):
+        """Directly extract features from the backbone+neck."""
+        x = self.backbone(img)
+        if self.with_neck:
+            x = self.neck(x)
+        return x
+
+    def forward_dummy(self, img):
+        """Used for computing network flops.
+
+        See `mmdetection/tools/get_flops.py`
+        """
+        raise NotImplementedError(
+            f'`forward_dummy` is not implemented in {self.__class__.__name__}')
+
+    def forward_train(self,
+                      img,
+                      img_metas,
+                      gt_bboxes,
+                      gt_labels,
+                      gt_bboxes_ignore=None,
+                      gt_masks=None,
+                      **kwargs):
+        """
+        Args:
+            img (Tensor): Input images of shape (B, C, H, W).
+                Typically these should be mean centered and std scaled.
+            img_metas (list[dict]): A List of image info dict where each dict
+                has: 'img_shape', 'scale_factor', 'flip', and may also contain
+                'filename', 'ori_shape', 'pad_shape', and 'img_norm_cfg'.
+                For details on the values of these keys see
+                :class:`mmdet.datasets.pipelines.Collect`.
+            gt_bboxes (list[Tensor]): Each item are the truth boxes for each
+                image in [tl_x, tl_y, br_x, br_y] format.
+            gt_labels (list[Tensor]): Class indices corresponding to each box
+            gt_bboxes_ignore (None | list[Tensor]): Specify which bounding
+                boxes can be ignored when computing the loss.
+            gt_masks (list[:obj:`BitmapMasks`]) : The segmentation
+                masks for each box.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+
+        gt_masks = [
+            gt_mask.to_tensor(dtype=torch.bool, device=img.device)
+            for gt_mask in gt_masks
+        ]
+        x = self.extract_feat(img)
+        losses = dict()
+
+        # CondInst, yolact
+        if self.bbox_head:
+            # bbox_head_results is a tuple
+            bbox_head_preds = self.bbox_head(x)
+            # positive_infos is a obj:`DetectionResults`
+            # It contains the information about the positive samples
+            # CondInst, Yolact
+            det_losses, positive_infos = self.bbox_head.loss(
+                *bbox_head_preds,
+                gt_bboxes=gt_bboxes,
+                gt_labels=gt_labels,
+                img_metas=img_metas,
+                gt_bboxes_ignore=gt_bboxes_ignore)
+            losses.update(det_losses)
+        else:
+            positive_infos = None
+
+        mask_head_inputs = (x, gt_labels, gt_masks, img_metas)
+
+        # when no positive_infos add gt bbox
+        mask_loss = self.mask_head.forward_train(
+            *mask_head_inputs,
+            positive_infos=positive_infos,
+            gt_bboxes=gt_bboxes,
```

Comment:
Why are `gt_labels, gt_masks, img_metas` packed in the tuple while gt_bboxes and other parameters not?

Reply:
Compared to  `gt_bboxes, positive_infos`,  `gt_labels, gt_masks, img_metas` is essential 
arguments for all algorithms.